modules/internal/ChapelDynamicLoading.chpl

```diff
@@ -0,0 +1,797 @@
+
+          if numErrs.read() {
+            // If there were errors, then report the first error.
+            // TODO: Consolidate the reported errors.
+            for i in 0..<errBuf.size {
+              if errBuf[i] {
+                err = errBuf[i];
+                break;
+              }
+            }
+          } else {
+            // Otherwise, try to add the newly created info to the store on
+            // LOCALE-0, or give up if another task already beat us to it.
+
+            for i in 0..<bin._systemPtrs.size {
+              // Make sure that all locales have succeeded in 'dlopen()'.
+              assert(bin._systemPtrs[i] != nil);
+            }
+
+            manage store.withWriteLock() {
+              var found = store.get(bin._systemPtrs[0], ret);
+              if !found {
+                ret = owned.release(bin);
+                var ok = store.set(ret!._systemPtrs[0], ret);
+                assert(ok);
+              }
+            }
+          }
+        }
+      }
+
+      return ret;
+    }
+
+    inline proc type _systemDynLibClearError() do dlerror();
+
+    inline proc type _systemDynLibClose(ptr: c_ptr(void)) do dlclose(ptr);
+
+    inline proc type _systemDynLibWrapError() {
+      const sysErr = dlerror();
+      if sysErr != nil then
+        return new DynLibError?(try! string.createCopyingBuffer(sysErr));
+      return nil;
+    }
+
+    inline proc type
+    _systemDynLibOpen(path: string, out err: owned DynLibError?) {
+      _systemDynLibClearError();
+
+      var ret = dlopen(path.c_str(), RTLD_LAZY);
+      if ret == nil {
+        var tmp = _systemDynLibWrapError();
+        if tmp != nil then err = tmp;
+      }
+
+      return ret;
+    }
+
+    inline proc type
+    _systemDynLibLookup(sym: string, handle: c_ptr(void),
+                        out err: owned DynLibError?) {
+      _systemDynLibClearError();
+
+      var ret = dlsym(handle, sym.c_str());
+      if ret == nil {
+        var tmp = _systemDynLibWrapError();
+        if tmp != nil then err = tmp;
+      }
+
+      return ret;
+    }
+
+    // TODO: Also need to evict pointer cache entries.
+    proc _close() {
+      assert(_refCount.read() == 0);
+      if _closed then return;
+
+      on Locales[0] do manage this.withWriteLock() {
+        if !_closed {
+          _closed = true;
+          for i in 0..<_systemPtrs.size {
+            const loc = Locales[i];
+            const ptr = _systemPtrs[i];
+            if ptr != nil then on loc {
+              this.type._systemDynLibClose(ptr);
+            }
+          }
+        }
+      }
+    }
+
+    inline proc bumpRefCount() do _refCount.fetchAdd(1);
+    inline proc dropRefCount() {
+      if _refCount.read() <= 0 then return;
+      _refCount.fetchSub(1);
+    }
+
+    // TODO: 'T' must be a procedure type, but we cannot restrict it yet.
+    // Could add a primitive "any procedure type" to use instead.
+    proc loadSymbol(sym: string, type t, out err: owned DynLibError?) {
```

Comment:
can resolve the todo with a where clause?

Reply:
nevermind, 3 lines down you gave a check using isProcedure. So is the todo still applicable?

Reply:
The TODO is applicable. I am checking if it is in the body of the function but that is not the same as restricting it in the signature e.g., `type t: proc`. I've adjusted the TODO to mention the "any proc type constructor".